### DISCUSSION TITLE
Fabric Deep Dive

### DISCUSSION THREAD

**Discussion by cortinico** [reactwg/react-native-new-architecture#1] [Deep Dive]
Hi all,
I'm opening this discussion to discuss about one of the component of the New React Native Architecture: **the Fabric Renderer**.

Feel free to ask your questions/ideas/doubts related to this Fabric in this discussion.

## Relevant resources

- [Fabric Renderer on reactnative.dev](https://reactnative.dev/docs/next/fabric-renderer)
- [Fabric: Render, Commit and Mount](https://reactnative.dev/docs/next/render-pipeline)
- [Fabric: Cross Platform Implentation](https://reactnative.dev/docs/next/xplat-implementation)
- [Fabric: View Flattening](https://reactnative.dev/docs/next/view-flattening)
- [Fabric: Threading Model](https://reactnative.dev/docs/next/threading-model)
- [Enabling Fabric on Android](https://reactnative.dev/docs/next/new-architecture-app-renderer-android)
- [Enabling Fabric on iOS](https://reactnative.dev/docs/next/new-architecture-app-renderer-ios)
- https://github.com/react-native-community/discussions-and-proposals/issues/4

**Comment by AndreiCalazans** [comment]
> The operations are synchronously executed on UI thread. If the commit phase executes on background thread, the mounting phase is scheduled for the next “tick” of UI thread. On the other hand, if the commit phase executes on UI thread, mounting phase executes synchronously on the same thread.
 (i) From [The Render Pipeline](https://reactnative.dev/docs/next/render-pipeline)

When I first read the above it was not clear why there is this alternation of threads happening with the commit and mounting phase. It only made sense after reading [The Threading Model Post](https://reactnative.dev/docs/next/threading-model#render-in-the-ui-thread). Maybe it would be worth linking it there and introducing how priority updates impact the render pipeline?

**Reply by gorhom** [reply]
Also in the commit phase, it is says that `These operations are asynchronously executed on a background thread.` but we know that is not the case for all rendering scenarios.

**Comment by AndreiCalazans** [comment]
Small nit, but a quality of life improvement, it would be for [The Render Pipeline](https://reactnative.dev/docs/next/render-pipeline) post to have linkable headders to each phase:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/20777666/155977788-fb3c1ca2-9da4-4d96-a606-3cee2373a66a.png">

I only understood that section under the image was related to one of the rendering phases after the second one I think.

**Reply by cortinico** [reply]
Addressed here:
- https://github.com/facebook/react-native-website/pull/2985

**Reply by gorhom** [reply]
@cortinico i think this could be applied for [React State Updates](https://reactnative.dev/docs/next/render-pipeline#react-state-updates) too

**Reply by Simek** [reply]
>  i think this could be applied for [React State Updates](https://reactnative.dev/docs/next/render-pipeline#react-state-updates) too

Yup, agreed. PR addressing this is here:
* https://github.com/facebook/react-native-website/pull/2997

**Comment by AndreiCalazans** [comment]
About the [View Flattening post](https://reactnative.dev/docs/next/view-flattening). 

From following along I interpreted that [View Flattening is skipped in some cases](https://github.com/facebook/react-native/blob/1953f6f02e9b1b15498dff33badcb4ed2bcf9ca0/ReactCommon/react/renderer/components/view/ViewShadowNode.cpp#L66-L96), could we expand on that or document it in the post? specially since an app that unknownly causes View flattenning to not occur could have worse peformance.

**Reply by sammy-SC** [reply]
Hello @AndreiCalazans

it is always hard to know how much complexity to expose to engineers and what to hide. I didn't want to get into all of the implementation details in the deep dive as there are quite a few things that affect whether a node is flattened or not. Also, this optimisation should come for "free" and people shouldn't have to think about it.

Do you think we should document every aspect that affects view flattening? Or do you think a link to code would be enough?

**Reply by AndreiCalazans** [reply]
> it is always hard to know how much complexity to expose to engineers and what to hide.  

That's a fair point. 

I think just the link to the code might be enough since this is a deep dive and not targetted at the novice developer.

**Reply by jacobp100** [reply]
Something really love to see is a `<Layout>` component: a subset of `<View>` which - either via typechecking or runtime checks - ensures the view will be flattened.

**Reply by sammy-SC** [reply]
> Something really love to see is a `<Layout>` component: a subset of `<View>` which - either via typechecking or runtime checks - ensures the view will be flattened.

Interesting idea @jacobp100. I'll share with you my thoughts on the subject so we can discuss it further.

**Increased complexity** is my concern with `<Layout>`. I've always seen as one of the biggest strengths of React the level of abstraction it offers and how it papers over underlaying platforms. This makes it easier for newcomers to get started with React/React Native and the lower cognitive load is helpful for seasoned developers as well. 
Now, imagine we introduce a new component `<Layout>`. We will have to explain why it exists. That sounds like a trivial thing, but if you are getting started with programming, memory and CPU cycles are not straightforward concepts. View flattening is powerful because developers do not need to think about it, it happens automatically.

**Negligible performance gains** from using few `<Layout>`. To see any meaningful impact, we would have to use tens, if not hundreds of `<Layout>` components. Complexity of view hierarchy and underlaying CPU play a role in performance boost as well. But the big gain from view flattening comes if there are many views that get flattened. 

**Alternative approaches** we could explore before considering `<Layout>`. Right now, view flattening happens automatically and it is not exposed to developers in any way. I think we could explore a way to expose when view is flattened or not to developers and explain why that is happening. We could leverage existing tools like Flipper or React DevTools. This would incrementally expose developers to the complexity and give them information to make decision. 

What do you think @jacobp100 ?

**Comment by AndreiCalazans** [comment]
In [the discrete event interruption](https://reactnative.dev/docs/next/threading-model#discrete-event-interruption).

When it states that "React and the renderer are able to interrupt the render phase and merge its state with a high priority event that was executed on the UI thread." Does it mean the rendering that was happening in the JS thread will continue in the UI thread alongside the high priority event?

**Reply by sammy-SC** [reply]
Once discrete event has been handled, React will continue the render phase in JS thread. Only updates relevant to the discrete event will be applied synchronously on the main thread. This is to minimise time it blocks the main thread . I'll change the diagram so it clearer, thanks!

**Reply by sammy-SC** [reply]
Addressed: https://github.com/facebook/react-native-website/pull/2999

**Reply by JoonDong2** [reply]
@sammy-SC Just by looking at the image, it looks like the React tree is being modified on the UI thread.

Even if it is a "Discrete event", is it correct that a React trigger is created in the JS thread and reflected synchronously `to` the UI thread instead of being scheduled for the next "tick"?

**Comment by AndreiCalazans** [comment]
Nit: [In the architecture overview](https://reactnative.dev/docs/next/architecture-overview) it would be nice to point out how knowing the architecture is not a pre-requisite to being a good React Native developer. Knowing how the technical recruiting world works and how well detailed the deep-dive is it will not be a surprise if technical interviews start to ask questions related to the inner workings of React Native. We shoud try to discourage that behavior.

**Reply by sammy-SC** [reply]
That's a great point! I'll try to include it in the introduction. Thank you.

**Reply by sammy-SC** [reply]
Hello @AndreiCalazans,

I forgot to follow up on this. I've updated the introduction section in [architecture overview](https://reactnative.dev/architecture/overview) to clarify who is the target audience and that it is not a pre-requisite to be a good React Native developer. It now also redirects people to Guides section in case newcomers stumble there. 
This should address the issue you pointed out. 

Thank you for your feedback.

**Comment by gorhom** [comment]
in [Fabric Renderer](https://reactnative.dev/docs/next/fabric-renderer)

>This documentation provides an overview of the [new renderer](https://reactnative.dev/docs/next/architecture-glossary#fabric-render) and its concepts. It avoids platform specifics and doesn’t contain any code snippets or pointers.

nit: the link is not correct in `new renderer`, it should point to https://reactnative.dev/docs/next/architecture-glossary#fabric-renderer

**Reply by cortinico** [reply]
Just checked and the link looks fine to me (it points to where you mentioned).

**Comment by gorhom** [comment]
in [Fabric Renderer](https://reactnative.dev/docs/next/fabric-renderer)

>Type safety: code generation to ensure type safety across the JS and [host platforms](https://reactnative.dev/docs/next/architecture-glossary#host-platform). 

i think we should add a page/documentation for the `code generation`, if there isn't already.

**Reply by sammy-SC** [reply]
Hello @gorhom,

do you suggest adding a deep dive on code generation? Or adding documentation how to use code generation in your app.

If it is the former, is there anything specific you would be interested in? 

If it is the latter, then I believe there is a plan to add documentation for code generation to the [new architecture migration guide playbook](https://reactnative.dev/docs/next/new-architecture-intro). @cortinico will know more about this.

**Reply by cortinico** [reply]
> If it is the latter, then I believe there is a plan to add documentation for code generation to the [new architecture migration guide playbook](https://reactnative.dev/docs/next/new-architecture-intro). @cortinico will know more about this.

Yes. Currently there are details about the Codegen in the New Architecture Guide:
https://reactnative.dev/docs/new-architecture-library-ios
https://reactnative.dev/docs/new-architecture-library-android

As a side note, we're currently expanding this guide and restructuring it a bit (@cipolleschi is driving this effort). We specifically planned for a page only about Codegen to document the rationale behind this component and give specific insights about it.

https://github.com/facebook/react-native-website/pull/3037/files#diff-47fbdf8e212fcd2636a90f2e0a70528f7bd0a0a68da28aea8d790250fb334ede

Ideally this would clarify a bit more your doubts @gorhom

**Reply by gorhom** [reply]
hi @sammy-SC , @cortinico 👋

I meant the former, "adding documentation how to use code generation in your app or with library".

>Yes. Currently there are details about the Codegen in the New Architecture Guide:
https://reactnative.dev/docs/new-architecture-library-ios
https://reactnative.dev/docs/new-architecture-library-android

this is exactly what i was referencing and looking for 👏

> As a side note, we're currently expanding this guide and restructuring it a bit (@cipolleschi is driving this effort). We specifically planned for a page only about Codegen to document the rationale behind this component and give specific insights about it.
> https://github.com/facebook/react-native-website/pull/3037/files#diff-47fbdf8e212fcd2636a90f2e0a70528f7bd0a0a68da28aea8d790250fb334ede

this would be perfect! thanks for the effort spent on these documents <3

**Comment by gorhom** [comment]
in [Render Pipeline](https://reactnative.dev/docs/next/render-pipeline#initial-render)

>React recursively reduces this React Element to a terminal [React Host Component](https://reactnative.dev/docs/next/architecture-glossary#host-view-tree-and-host-view)

nit: the link is pointing to `Host View`, but it should be https://reactnative.dev/docs/next/architecture-glossary#react-host-components-or-host-components

**Reply by cortinico** [reply]
Fixed here https://github.com/facebook/react-native-website/pull/2992

**Comment by gorhom** [comment]
in [Render Pipeline > React State Update](https://reactnative.dev/docs/next/render-pipeline#react-state-updates)

>Layout Calculation: Similar to Layout Calculation during [Initial Render](https://reactnative.dev/docs/next/render-pipeline#initial-render). One important difference is that layout calculation may cause shared React Shadow Nodes to be cloned.

I think it would be clear/helpful to reference the previous example cloned node : `may cause shared React Shadow Nodes (Node 4 in previous example) to be cloned.`

**Comment by bozaigao** [comment]
@cortinico I was confused about the rendering performance of fabric. I wrote a demo to test the rendering performance of fabric. I rendered 1000 TestComponent components in scrollview. The previous JS Bridge architecture took more than 700 milliseconds ,how ever Fabric takes more than 1400 milliseconds, rendering performance of fabric is worse than before? I am very confused,repo url:https://github.com/bozaigao/FabricDemo

**Reply by chakrihacker** [reply]
I am not able to check the code

**Reply by sammy-SC** [reply]
Hello @bozaigao,

I would be curious to take a look at this, could you please fix the repo? What platform was this on?

**Comment by billp72** [comment]
I'm using this repo (
https://github.com/talknagish/react-native-turbo-starter) to test out
Fabric but am experiencing issues suggesting some native components are
32bit and need to be 64bit so won't build because the JVM is running out of
memory.

Before I go crazy trying to fix this, I tried:
//gradle.properties
org.gradle.daemon=true
org.gradle.jvmargs=-Xmx4096m

and
//app/build.gradle
defaultConfig {
    ndk.abiFilters 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'
    ...
}

any other suggestions?
thanks

**Reply by chakrihacker** [reply]
Curious what are you building that needs 64bit number??

**Reply by billp72** [reply]
Nothing, actually. I haven't made any modifications to the code in this respect. The repo I cloned must be set (somewhere) to use 64bit. I'd be fine with changing to 32bit to get it to work. Thanks

**Comment by billp72** [comment]
I haven't made any modifications to the existing repo. I cloned it, ran
yarn install, on my windows machine, then ran yarn android just as it said.

The first error said it couldn't find a specific cmake version in the SDK
path. So I added cmake.dir to local.properties and pointed it to the
missing version as suggested.

I tried running yarn android again and this time it stalled at
app:something. Eventually, the jvm ran out of memory. That's when I saw the
message (paraphrasing) 'native modules need to run 64bit'.

I googled around, hoping their might be a quick fix but so far no luck.

Thanks.

**Comment by JoonDong2** [comment]
It was mentioned that the render phase is the phase that creates the React tree and shadow tree.
[https://reactnative.dev/architecture/render-pipeline](https://reactnative.dev/architecture/render-pipeline)

However, in ["Discrete event interruption",](https://reactnative.dev/architecture/threading-model#discrete-event-interruption) section it was mentioned that the render stage may run on the UI thread. 

In what cases can the React tree be reconciled on the UI thread? Or does the render step in the UI thread only mean changing the shadow tree?
